### PR TITLE
Sign up in steps

### PR DESCRIPTION
**Pull request by santiagopais**
Changed sign up to do it in steps, setting the password after confirming the email

### CODE REVIEW

File: app/controllers/users/confirmations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+module Users
+  class ConfirmationsController < Devise::ConfirmationsController
+    # skip_before_action :require_no_authentication
+    # skip_before_action :authenticate_user!
+
+    # Se ejecuta luego de tocar confirm account, lo sobreescribo para
```
Comment (martin8767) [inline app/controllers/users/confirmations_controller.rb:8]: remove this comments

File: app/models/user.rb [hunk 2]
```diff
@@ -7,6 +7,53 @@ class User < ApplicationRecord
...
   validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }, allow_blank: true
-  validates :username, presence: true, length: { minimum: 3 }, uniqueness: { case_sensitive: true }
-  validates :encrypted_password, presence: true
+  validates :username, presence: true, length: { in: 2..20 }, uniqueness: { case_sensitive: true }
+  validates :encrypted_password, presence: true, if: :password_required?
+  validates :first_name, presence: true, length: { minimum: 2 }
+  validates :bio, length: { maximum: 160 }
+  validates :website, format: { with: %r{https?://(www.)?[^\W]*\.com} }, allow_blank: true
+  validates :birthdate, comparison: { less_than: 18.years.ago }
+  # validates :password, :password_confirmation, presence: true, if: :password_required?
```
Comment (martin8767) [inline app/models/user.rb:16]: remove this

File: app/models/user.rb [hunk 3]
```diff
@@ -7,6 +7,53 @@ class User < ApplicationRecord
...
+  end
+
+  def only_if_unconfirmed(&)
+    pending_any_confirmation(&) # Checks whether the record requires any confirmation.
+  end
+
+  def password_required?
+    # Password is required if it is being set, but not for new records
+    if persisted? # returns true if it's not a new record and it was  not destroyed, otherwise returns false.
+      !password.nil? || !password_confirmation.nil? # Si el record no es nuevo, retorna true solo si la contraseña o la confirmación no es vacia.
```
Comment (martin8767) [inline app/models/user.rb:46]: we don't need this comments, the code is self explanatory

File: app/views/devise/confirmations/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+<div class="[background:#FFF] mx-auto my-32">
+  <div class="flex flex-col items-center justify-center w-[600px] h-[650px] shrink-0 [background:#FFF] shadow-[0px_0px_20px_0px_rgba(0,0,0,0.15)] rounded-[29px]">
+    <!--h3 class="heading-3 m-6">Step 2 of 3</h3> -->
```
Comment (martin8767) [inline app/views/devise/confirmations/show.html.erb:3]: remove

File: app/models/user.rb [hunk 5]
```diff
@@ -7,6 +7,52 @@ class User < ApplicationRecord
...
+    update(p)
+  end
+
+  # new function to return whether a password has been set
+  def no_password?
+    encrypted_password.blank?
+  end
+
+  def only_if_unconfirmed(&)
+    pending_any_confirmation(&) # Checks whether the record requires any confirmation.
```
[the remark this answers is not in the document]
Comment (santiagopais) [reply app/models/user.rb:43]: pending_any_confirmation is a protected method of confirmable. I create this method to use it in the controller

**Pull request merged**